src/hardware/sipeed-slogic-analyzer/api.c

```diff
@@ -0,0 +1,711 @@
+		do {
+			slogic_usb_control_read(sdi, SLOGIC16U3_CONTROL_IN_REQ_REG_READ, SLOGIC16U3_R32_AUX, 0x0000, cmd_aux, 4, 500);
+			sr_dbg("[%u]read aux samplerate: %08x.", retry, ((uint32_t*)cmd_aux)[0]);
+			retry += 1;
+			if (retry > 5)
+				return SR_ERR_TIMEOUT;
+		} while (!(cmd_aux[2] & 0x01));
+		sr_dbg("samplerate length: %u.", (*(uint16_t*)cmd_aux)>>9);
+
+		while (((uint16_t*)(cmd_aux+4))[0] <= 1) {
+			slogic_usb_control_read(sdi, SLOGIC16U3_CONTROL_IN_REQ_REG_READ, SLOGIC16U3_R32_AUX + 4, 0x0000, cmd_aux + 4, (*(uint16_t*)cmd_aux)>>9, 500);
+
+			sr_dbg("aux: %u %u %u %u %x %u %u.", cmd_aux[0], cmd_aux[1], cmd_aux[2], cmd_aux[3], ((uint16_t*)(cmd_aux+4))[0], ((uint16_t*)(cmd_aux+4))[1], ((uint32_t*)(cmd_aux+4))[1]);
+
+
+			uint64_t base = SR_MHZ(1) * ((uint16_t*)(cmd_aux+4))[1];
+			if (base % devc->cur_samplerate) {
+				sr_dbg("Failed to configure samplerate from base[%u] %u.", ((uint16_t*)(cmd_aux+4))[0], base);
+				((uint16_t*)(cmd_aux+4))[0] += 1;
+				slogic_usb_control_write(sdi, SLOGIC16U3_CONTROL_OUT_REQ_REG_WRITE, SLOGIC16U3_R32_AUX + 4, 0x0000, cmd_aux + 4, 4, 500);
+				continue;
+			}
+			uint32_t div = base / devc->cur_samplerate;
+			((uint32_t*)(cmd_aux+4))[1] = div;
+
+			sr_dbg("aux: %u %u %u %u %x %u %u.", cmd_aux[0], cmd_aux[1], cmd_aux[2], cmd_aux[3], ((uint16_t*)(cmd_aux+4))[0], ((uint16_t*)(cmd_aux+4))[1], ((uint32_t*)(cmd_aux+4))[1]);
+			slogic_usb_control_write(sdi, SLOGIC16U3_CONTROL_OUT_REQ_REG_WRITE, SLOGIC16U3_R32_AUX + 4, 0x0000, cmd_aux + 4, (*(uint16_t*)cmd_aux)>>9, 500);
+
+			slogic_usb_control_read(sdi, SLOGIC16U3_CONTROL_IN_REQ_REG_READ, SLOGIC16U3_R32_AUX + 4, 0x0000, cmd_aux + 4, (*(uint16_t*)cmd_aux)>>9, 500);
+			sr_dbg("aux: %u %u %u %u %x %u %u.", cmd_aux[0], cmd_aux[1], cmd_aux[2], cmd_aux[3], ((uint16_t*)(cmd_aux+4))[0], ((uint16_t*)(cmd_aux+4))[1], ((uint32_t*)(cmd_aux+4))[1]);
+			break;
+		}
+
+		if (((uint16_t*)(cmd_aux+4))[0] <= 1) {
+			sr_dbg("Succeed to configure samplerate.");
+		} else {
+			sr_dbg("Failed to configure samplerate.");
+		}
+	}
+
+	return slogic_usb_control_write(sdi, SLOGIC16U3_CONTROL_OUT_REQ_REG_WRITE, SLOGIC16U3_R32_CTRL, 0x0000, ARRAY_AND_SIZE(cmd_run), 500);
+}
+
+static int slogic16U3_remote_stop(const struct sr_dev_inst *sdi) {
+	const uint8_t cmd_rst[] = {0x02, 0x00, 0x00, 0x00};
+	return slogic_usb_control_write(sdi, SLOGIC16U3_CONTROL_OUT_REQ_REG_WRITE, SLOGIC16U3_R32_CTRL, 0x0000, ARRAY_AND_SIZE(cmd_rst), 500);
+}
+/* SLogic16U3 end */
+
+static int slogic_usb_control_write(const struct sr_dev_inst *sdi, uint8_t request, uint16_t value, uint16_t index, uint8_t *data, size_t len, int timeout)
+{
+	int ret;
+	struct dev_context *devc;
+	struct sr_usb_dev_inst *usb;
+
+	devc = sdi->priv;
+	usb  = sdi->conn;
+
+	sr_spew("%s: req:%u value:%u index:%u %p:%u in %dms.", __func__, request, value, index, data, len, timeout);
+	if (!data && len) {
+		sr_warn("%s: Nothing to write although len(%u)>0!", __func__, len);
+		len = 0;
+	} else if (len & 0x3) {
```

Comment:
Nope, this has to be a hard error, you will do an out-of-bounds read (or even write below).